x/twap/logic.go

```diff
@@ -225,14 +229,17 @@ func (k Keeper) getMostRecentRecord(ctx sdk.Context, poolId uint64, assetA, asse
 // computeArithmeticTwap computes and returns an arithmetic TWAP between
 // two records given the quote asset.
 // precondition: endRecord.Time >= startRecord.Time
-// if endRecord.LastErrorTime is after startRecord.Time, return an error at end + result
+// if (endRecord.LastErrorTime >= startRecord.Time) returns an error at end + result
+// if (startRecord.LastErrorTime == startRecord.Time) returns an error at end + result
 // if (endRecord.Time == startRecord.Time) returns endRecord.LastSpotPrice
 // else returns
 // (endRecord.Accumulator - startRecord.Accumulator) / (endRecord.Time - startRecord.Time)
 func computeArithmeticTwap(startRecord types.TwapRecord, endRecord types.TwapRecord, quoteAsset string) (sdk.Dec, error) {
 	// see if we need to return an error, due to spot price issues
 	var err error = nil
-	if endRecord.LastErrorTime.After(startRecord.Time) || endRecord.LastErrorTime.Equal(startRecord.Time) {
+	if endRecord.LastErrorTime.After(startRecord.Time) ||
+		endRecord.LastErrorTime.Equal(startRecord.Time) ||
+		startRecord.LastErrorTime.Equal(startRecord.Time) {
```

Comment:
curious, do we not need to add `startRecord.LastErrorTime.Equal(startRecord.Time)`?

Reply:
Consider the case where:
- `endRecord = {Time: 2, LastErrTime: 0}`
- `startRecord = {Time: 1, LastErrTime: 1}`

The conditions will evaluate to the following:
- `endRecord.LastErrorTime.After(startRecord.Time)` - False
- `endRecord.LastErrorTime.Equal(startRecord.Time)` - False
- `startRecord.LastErrorTime.Equal(startRecord.Time)` - True

Without the third condition, the calculation proceeds. And we do not want it to proceed because the startRecord has an error at its time.

The case is definitely possible because the startRecord might have been interpolated before being passed to this function (e.g. from `{Time: 0, LastErrTime: 0}` to `{Time: 1, LastErrTime: 1}`) while the endRecord is untouched.